A32-xds-circuit-breaking.md

```diff
@@ -0,0 +1,157 @@
+many circuit breaking limits are configured for connection/connection pool 
+management. Since each gRPC client creates at most one HTTP/2 connection to 
+a single upstream endpoint address, those configurations do not apply to gRPC
+and will be ignored. See more in [Other configurations considered](#other-configurations-considered).
+
+As per upstream cluster configurations, circuit breakers will be implemented as
+part of the CDS LB policy in [gRPC client's xDS architecture](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md).
+After the [RouteAction](https://github.com/grpc/proposal/pull/192) is 
+implemented for [traffic splitting and routing](https://github.com/grpc/proposal/blob/master/A28-xds-traffic-splitting-and-routing.md),
+each CDS LB policy uniquely represents the load balancing logic for all requests
+sent to the upstream cluster.
+
+## Overview
+
+gRPC client receives circuit breaking configurations as part of the [Cluster](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster.proto#cluster) 
+resource configuration in Cluster Discovery Service (CDS) responses. Each
+circuit breaker [Thresholds](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/cluster/circuit_breaker.proto#cluster-circuitbreakers-thresholds)
+defines settings for a RoutingPriority. gRPC only takes the first 
+_Thresholds_ with `priority` being [`DEFABUT`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/base.proto#enum-core-routingpriority).
+
+As of the current gRPC design, only the `max_requests` field, which limits the 
+maximum number of requests that can be outstanding to all hosts in a cluster 
+at any given time, in `Thresholds` is considered. All the others are either 
+inapplicable to gRPC or the corresponding features are missing for now. The
+design and implementation section of this proposal mainly talks about the
+enforcement of the limit for maximum number of outstanding requests sent to
+an upstream cluster. See [Other configurations considered](#other-configurations-considered) 
+for rational.
+
+To match Envoy's implementation, circuit breaking limits will always be 
+enforced with certain defaults, even if the gRPC client receives no applicable 
+circuit breakers. Users can effectively turn circuit breaking off by setting
+thresholds to very high values, for example, to 
+`std::numeric_limits<uint32_t>::max()`.
+
+## Detailed design
+
+### Maintaining thresholds
+
+Each CDS LB policy maintains a threshold for the maximum number of outstanding 
+requests to hosts in the cluster that this policy is load balancing for. Make
+sure this threshold can be updated when receiving an update from the
+XdsClient cluster watch interface and the updated value can be dynamically 
+picked up by request limiting logic. By default, this value is set to 1024.
+
+### Limiting outstanding requests
+
+Each CDS LB policy uses a counter to count the number of requests sent through
```

Comment:
I think we should write this a little more generally, without dictating quite so much of the implementation details.  I suggest something like the following:

The CDS policy's picker will enforce the configured limit.  Each CDS LB policy will maintain a counter of the number of requests currently in flight to the associated cluster.  The counter will be incremented when a request is started on a subchannel and decremented when the request is complete.  When the picker is going to start a call, it will check whether the current number of in-flight requests is lower than the configured limit.  If so, the picker will do the pick as usual, returning a subchannel.  If not, it will fail the call with status UNAVAILABLE.

Reply:
Rephrased.